Switch to esbuild and tsc instead of pika

kfcampbell opened the pull request:
Replicating [octokit/request.js#584](https://github.com/octokit/request.js/pull/584) to another repo. 

---

Pika has been deprecated for a while now, and the project has now been archived, and because of that the released npm package is missing most of the files generated by the build step.

Uses esbuild to transpile the TS source code into an ESM source, NodeJs bundle, and a browser bundle

Uses the TypeScript compiler to generate the types.

## Behavior

### Before the change?
<!-- Please describe the current behavior that you are modifying. -->

* Uses pika for the build system
* The released npm package is missing most of the files generated by the build step. `dist-node`, `dist-types`, `dist-web`... even though they are generated correctly.

### After the change?
<!-- Please describe the behavior or changes that are being added by this PR. -->

* Uses esbuild to generate the JS bundles
* Uses the TypeScript compiler to generate the type definitions
* The correct file matching patterns are used so we publish all the necessary files again.


----

## Additional info

### Pull request checklist
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Added the appropriate label for the given change

### Does this introduce a breaking change?
<!-- If this introduces a breaking change make sure to note it here any what the impact might be -->

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/main/community/breaking_changes.md) to help!

- [ ] Yes (Please add the `Type: Breaking change` label)
- [x] No

If `Yes`, what's the impact:

* N/A


### Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. -->
<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

Please add the corresponding label for change this PR introduces:
- Bugfix: `Type: Bug`
- Dependencies/code cleanup: `Type: Maintenance`

----

wolfy1339 commented:
Don't forget to explicitly mark type imports and exports.

Whenever an imported value, that is also JS code, is only used as a type, it should be imported as a type, to make sure that it doesn't show up in the build output.

All typescript types (except for `enum`) need to be imported using type imports, and exports need to use the type export syntax.

kfcampbell commented:
@wolfy1339 I've updated this PR to correctly import types, bumped to the latest version of octokit/tsconfig, and fixed the merge conflict! Hopefully this is good to go now.

kfcampbell commented:
Hmm...locally I'm seeing:

![image](https://github.com/octokit/plugin-throttling.js/assets/9327688/1f493011-d756-4019-9f77-00c837b3f878)

and:

![image](https://github.com/octokit/plugin-throttling.js/assets/9327688/dac72640-6331-432b-84b4-3fba5e3c3bc8)

Even after many reloads here, I'm not seeing the new commits: 

![image](https://github.com/octokit/plugin-throttling.js/assets/9327688/6dd7a4fc-e0c8-4e00-8348-3dadcc429111)

I wonder if we're having issues.

kfcampbell commented:
Yep, looks like that's it:

![image](https://github.com/octokit/plugin-throttling.js/assets/9327688/717bf5fd-4654-4a36-8dbf-19aac1885d3f)

kfcampbell commented:
Looks like we're all good now!

kfcampbell commented on scripts/build.mjs:
```diff
@@ -0,0 +1,88 @@
+      outdir: "pkg/dist-node",
+      bundle: true,
+      platform: "node",
+      target: "node14",
```
Ahh, thanks! I missed that the build script also needed to be updated for the Node target.
